File: c/src/classical/substitution/polygraphic/alphabet_utils.c
/**
 * @file alphabet_utils.c
 * @brief Alphabet utilities implementation for polygraphic substitution ciphers
 */

#include "cryptology/classical/substitution/polygraphic/alphabet_utils.h"
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <stdbool.h>

#define MAX_ALPHABET_SIZE 256
#define MAX_LANGUAGE_SIZE 32

// Turkish letter frequency data (approximate)
static const char* turkish_frequent_letters = "aeirnldkysbzgcfhjpvmotu";

// English letter frequency data (approximate)
static const char* english_frequent_letters = "etaoinshrdlcumwfgypbvkjxqz";

const char* detect_language(const char *alphabet) {
    if (!alphabet) {
        return "unknown";
    }
    
    // Check for Turkish indicators
    if (strstr(alphabet, "ç") || strstr(alphabet, "ğ") || strstr(alphabet, "ı") || 
        strstr(alphabet, "ö") || strstr(alphabet, "ş") || strstr(alphabet, "ü") ||
        strstr(alphabet, "Ç") || strstr(alphabet, "Ğ") || strstr(alphabet, "İ") || 
        strstr(alphabet, "Ö") || strstr(alphabet, "Ş") || strstr(alphabet, "Ü")) {
        return "turkish";
    }
    
    // Check for English (26 letters, all basic Latin)
    if (strlen(alphabet) == 26) {
        bool is_english = true;
        for (int i = 0; alphabet[i] != '\0'; i++) {
            char c = tolower((unsigned char)alphabet[i]);
            if (c < 'a' || c > 'z') {
                is_english = false;
                break;
            }
        }
        if (is_english) {
            return "english";
        }
    }
    
    return "unknown";
}

int get_square_size(int alphabet_length) {
    if (alphabet_length <= 0) {
        return 5; // Default to 5x5
    }
    
    return (int)ceil(sqrt((double)alphabet_length));
}

static int apply_turkish_combinations(const char *alphabet, char *result, size_t result_size) {
    if (!alphabet || !result || result_size == 0) {
        return -1;
    }
    
    size_t len = strlen(alphabet);
    if (len >= result_size) {
        return -1;
    }
    
    strcpy(result, alphabet);
    
    // Apply Turkish combination rules
    char *pos;
    while ((pos = strstr(result, "ç")) != NULL) {
        *pos = 'c';
    }
    while ((pos = strstr(result, "ğ")) != NULL) {
        *pos = 'g';
    }
    while ((pos = strstr(result, "ı")) != NULL) {
        *pos = 'i';
    }
    while ((pos = strstr(result, "ö")) != NULL) {
        *pos = 'o';
    }
    while ((pos = strstr(result, "ş")) != NULL) {
        *pos = 's';
    }
    while ((pos = strstr(result, "ü")) != NULL) {
        *pos = 'u';
    }
    
    // Handle uppercase versions
    while ((pos = strstr(result, "Ç")) != NULL) {
        *pos = 'C';
    }
    while ((pos = strstr(result, "Ğ")) != NULL) {
        *pos = 'G';
    }
    while ((pos = strstr(result, "İ")) != NULL) {
        *pos = 'I';
    }
    while ((pos = strstr(result, "Ö")) != NULL) {
        *pos = 'O';
    }
    while ((pos = strstr(result, "Ş")) != NULL) {
        *pos = 'S';
    }
    while ((pos = strstr(result, "Ü")) != NULL) {
        *pos = 'U';
    }
    
    return 0;
}

static int remove_duplicates(const char *alphabet, char *result, size_t result_size) {
    if (!alphabet || !result || result_size == 0) {
        return -1;
    }
    
    bool seen[256] = {false};
    size_t pos = 0;
    
    for (size_t i = 0; alphabet[i] != '\0' && pos < result_size - 1; i++) {
        unsigned char c = (unsigned char)alphabet[i];
        if (!seen[c]) {
            result[pos++] = alphabet[i];
            seen[c] = true;
        }
    }
    
    result[pos] = '\0';
    return 0;
}

static int select_by_frequency(const char *alphabet, const char *frequent_letters, 
                               int target_size, char *result, size_t result_size) {
    if (!alphabet || !frequent_letters || !result || result_size == 0) {
        return -1;
    }
    
    bool used[256] = {false};
    size_t pos = 0;
    
    // First, add letters from frequent_letters that exist in alphabet
    for (int i = 0; frequent_letters[i] != '\0' && pos < target_size && pos < result_size - 1; i++) {
        char c = frequent_letters[i];
        if (strchr(alphabet, c) && !used[(unsigned char)c]) {
            result[pos++] = c;
            used[(unsigned char)c] = true;
        }
    }
    
    // If we still need more letters, add remaining letters from alphabet
    for (size_t i = 0; alphabet[i] != '\0' && pos < target_size && pos < result_size - 1; i++) {
        char c = alphabet[i];
        if (!used[(unsigned char)c]) {
            result[pos++] = c;
            used[(unsigned char)c] = true;
        }
    }
    
    result[pos] = '\0';
    return 0;
}

int combine_similar_letters(const char *alphabet, const char *language,
                           char *result, size_t result_size) {
    if (!alphabet || !result || result_size == 0) {
        return -1;
    }
    
    char temp[MAX_ALPHABET_SIZE];
    char deduplicated[MAX_ALPHABET_SIZE];
    
    // Detect language if auto
    const char *detected_language = language;
    if (strcmp(language, "auto") == 0) {
        detected_language = detect_language(alphabet);
    }
    
    // Apply language-specific combinations
    if (strcmp(detected_language, "turkish") == 0) {
        if (apply_turkish_combinations(alphabet, temp, sizeof(temp)) != 0) {
            return -1;
        }
    } else {
        // For English or unknown, just copy
        strncpy(temp, alphabet, sizeof(temp) - 1);
        temp[sizeof(temp) - 1] = '\0';
    }
    
    // Remove duplicates
    if (remove_duplicates(temp, deduplicated, sizeof(deduplicated)) != 0) {
        return -1;
    }
    
    // If still too long, use frequency-based selection
    int target_size = 25; // For 5x5 square
    if (strlen(deduplicated) > target_size) {
        const char *frequent_letters = NULL;
        if (strcmp(detected_language, "turkish") == 0) {
            frequent_letters = turkish_frequent_letters;
        } else {
            frequent_letters = english_frequent_letters;
        }
        
        if (select_by_frequency(deduplicated, frequent_letters, target_size, result, result_size) != 0) {
            return -1;
        }
    } else {
        strncpy(result, deduplicated, result_size - 1);
        result[result_size - 1] = '\0';
    }
    
    return 0;
}

int create_square_alphabet(const char *alphabet, int square_size,
                          char *result, size_t result_size) {
    if (!alphabet || !result || result_size == 0 || square_size <= 0) {
        return -1;
    }
    
    int target_length = square_size * square_size;
    size_t alphabet_len = strlen(alphabet);
    
    if (alphabet_len == target_length) {
        strncpy(result, alphabet, result_size - 1);
        result[result_size - 1] = '\0';
        return 0;
    } else if (alphabet_len < target_length) {
        // Pad with X if too short
        strncpy(result, alphabet, result_size - 1);
        result[result_size - 1] = '\0';
        
        size_t current_len = strlen(result);
        for (int i = current_len; i < target_length && i < result_size - 1; i++) {
            result[i] = 'X';
        }
        result[target_length] = '\0';
        return 0;
    } else {
        // Truncate if too long
        strncpy(result, alphabet, target_length);
        result[target_length] = '\0';
        return 0;
    }
}

int create_caesared_alphabet(const char *base_alphabet, int shift,
                            char *result, size_t result_size) {
    if (!base_alphabet || !result || result_size == 0) {
        return -1;
    }
    
    size_t len = strlen(base_alphabet);
    if (len >= result_size) {
        return -1;
    }
    
    shift = shift % (int)len;
    if (shift < 0) {
        shift += len;
    }
    
    strcpy(result, base_alphabet + shift);
    strncat(result, base_alphabet, shift);
    result[result_size - 1] = '\0';
    
    return 0;
}

const char* get_letter_combination_rules(const char *language) {
    if (!language) {
        return NULL;
    }
    
    if (strcmp(language, "turkish") == 0) {
        return "ç→c, ğ→g, ı→i, ö→o, ş→s, ü→u";
    } else if (strcmp(language, "english") == 0) {
        return "No combination rules needed";
    }
    
    return "Unknown language";
}

File: python/cryptology/classical/substitution/polygraphic/alphabet_utils.py
"""
Alphabet utilities for polygraphic substitution ciphers.

This module provides utilities for handling custom alphabets in polygraphic ciphers,
including letter combination strategies and square size calculations.
"""

import re
from typing import List, Tuple, Optional


def get_square_size(alphabet_length: int) -> int:
    """
    Calculate the appropriate square size for a given alphabet length.
    
    Args:
        alphabet_length: Number of letters in the alphabet
        
    Returns:
        Square size (e.g., 5 for 25 letters, 6 for 36 letters)
    """
    import math
    return math.ceil(math.sqrt(alphabet_length))


def combine_similar_letters(alphabet: str, language: str = "auto") -> str:
    """
    Combine similar letters in an alphabet to fit polygraphic cipher requirements.
    
    Args:
        alphabet: Input alphabet
        language: Language hint for combination rules
        
    Returns:
        Alphabet with similar letters combined
    """
    if language == "auto":
        language = detect_language(alphabet)
    
    if language == "turkish":
        return _combine_turkish_letters(alphabet)
    elif language == "russian":
        return _combine_russian_letters(alphabet)
    elif language == "german":
        return _combine_german_letters(alphabet)
    else:
        return _combine_generic_letters(alphabet)


def _combine_turkish_letters(alphabet: str) -> str:
    """Combine Turkish letters for polygraphic ciphers."""
    combinations = {
        'ç': 'c', 'ğ': 'g', 'ı': 'i', 'ö': 'o', 'ş': 's', 'ü': 'u',
        'Ç': 'C', 'Ğ': 'G', 'İ': 'I', 'Ö': 'O', 'Ş': 'S', 'Ü': 'U'
    }
    
    result = alphabet
    for turkish, english in combinations.items():
        result = result.replace(turkish, english)
    
    # Remove duplicates while preserving order
    seen = set()
    unique_chars = []
    for char in result:
        if char not in seen:
            unique_chars.append(char)
            seen.add(char)
    
    return ''.join(unique_chars)


def _combine_russian_letters(alphabet: str) -> str:
    """Combine Russian letters for polygraphic ciphers."""
    combinations = {
        'ё': 'е', 'й': 'и', 'ъ': '', 'ь': '',
        'Ё': 'Е', 'Й': 'И', 'Ъ': '', 'Ь': ''
    }
    
    result = alphabet
    for russian, replacement in combinations.items():
        result = result.replace(russian, replacement)
    
    # Remove duplicates while preserving order
    seen = set()
    unique_chars = []
    for char in result:
        if char not in seen:
            unique_chars.append(char)
            seen.add(char)
    
    return ''.join(unique_chars)


def _combine_german_letters(alphabet: str) -> str:
    """Combine German letters for polygraphic ciphers."""
    combinations = {
        'ä': 'ae', 'ö': 'oe', 'ü': 'ue', 'ß': 'ss',
        'Ä': 'AE', 'Ö': 'OE', 'Ü': 'UE'
    }
    
    result = alphabet
    for german, replacement in combinations.items():
        result = result.replace(german, replacement)
    
    # Remove duplicates while preserving order
    seen = set()
    unique_chars = []
    for char in result:
        if char not in seen:
            unique_chars.append(char)
            seen.add(char)
    
    return ''.join(unique_chars)


def _combine_generic_letters(alphabet: str) -> str:
    """Generic letter combination for unknown languages."""
    # Remove duplicates while preserving order
    seen = set()
    unique_chars = []
    for char in alphabet:
        if char not in seen:
            unique_chars.append(char)
            seen.add(char)
    
    return ''.join(unique_chars)


def detect_language(alphabet: str) -> str:
    """
    Detect the language of an alphabet based on character patterns.
    
    Args:
        alphabet: Input alphabet
        
    Returns:
        Detected language ('turkish', 'russian', 'german', 'english', 'unknown')
    """
    alphabet_lower = alphabet.lower()
    
    # Turkish indicators
    if any(char in alphabet_lower for char in ['ç', 'ğ', 'ı', 'ö', 'ş', 'ü']):
        return "turkish"
    
    # Russian indicators
    if any(char in alphabet_lower for char in ['ё', 'й', 'ъ', 'ь']):
        return "russian"
    
    # German indicators
    if any(char in alphabet_lower for char in ['ä', 'ö', 'ü', 'ß']):
        return "german"
    
    # English indicators
    if all(char in 'abcdefghijklmnopqrstuvwxyz' for char in alphabet_lower):
        return "english"
    
    return "unknown"


def create_square_alphabet(alphabet: str, square_size: int) -> str:
    """
    Create a square-sized alphabet by combining letters if necessary.
    
    Args:
        alphabet: Input alphabet
        square_size: Desired square size
        
    Returns:
        Alphabet of exactly square_size² length
    """
    target_length = square_size * square_size
    
    if len(alphabet) == target_length:
        return alphabet
    elif len(alphabet) < target_length:
        # Pad with X if too short
        return alphabet + 'X' * (target_length - len(alphabet))
    else:
        # Truncate if too long
        return alphabet[:target_length]


def create_caesared_alphabet(base_alphabet: str, shift: int) -> str:
    """
    Create a 'Caesared' alphabet by shifting the base alphabet.
    
    Args:
        base_alphabet: Base alphabet to shift
        shift: Number of positions to shift
        
    Returns:
        Shifted alphabet
    """
    shift = shift % len(base_alphabet)
    return base_alphabet[shift:] + base_alphabet[:shift]


def get_letter_combination_rules() -> dict:
    """
    Get the letter combination rules for different languages.
    
    Returns:
        Dictionary mapping languages to their combination rules
    """
    return {
        "turkish": {
            "ç": "c", "ğ": "g", "ı": "i", "ö": "o", "ş": "s", "ü": "u",
            "Ç": "C", "Ğ": "G", "İ": "I", "Ö": "O", "Ş": "S", "Ü": "U"
        },
        "russian": {
            "ё": "е", "й": "и", "ъ": "", "ь": "",
            "Ё": "Е", "Й": "И", "Ъ": "", "Ь": ""
        },
        "german": {
            "ä": "ae", "ö": "oe", "ü": "ue", "ß": "ss",
            "Ä": "AE", "Ö": "OE", "Ü": "UE"
        }
    }
